### DISCUSSION TITLE
is it correct that all this 3 are sql injection safe? just for clarification

### DISCUSSION THREAD

**Discussion by ishiyamakaiyo** [prisma/prisma#20132] [Q&A] [answered]
```sql
 const A= await prisma.$queryRaw(Prisma.sql`select * from users where username=${username}::text`);
```
```sql
 const B= await prisma.$queryRaw`select * from users where username=${username}::text`;
```
```sql
   const C = await prisma.$queryRawUnsafe(
      'SELECT * FROM user WHERE users = $1',
      username)
```
meaning the above i don't have to manually do string escape or that is still necessary before it reaches the query? 
of course assuming the I validated length, data type, date formatting,email correct? 

while this is the actual unsafe version of executeRawUnsafe
```sql
const result = await prisma.$executeRawUnsafe(
  `select * from users where username=${username}`
)
```

**Comment by ludralph** [comment, marked as answer]
HI @ishiyamakaiyo 

```ts
const A = await prisma.$queryRaw(Prisma.sql`select * from users where username=${username}::text`);
```
For the query above, Prisma creates prepared statements that are safe from SQL injections.

```ts
const B = await prisma.$queryRaw`select * from users where username=${username}::text`;
```
For the query above, Prisma creates prepared statements that are safe from SQL injections.

```ts
const C = await prisma.$queryRawUnsafe('SELECT * FROM user WHERE users = $1', username);
```
For this one, you are using `$queryRawUnsafe`, `$queryRawUnsafe` supports standard parameterized queries where each variable is represented by a symbol (? for mySQL, $1, $2, and so on for PostgreSQL). As with tagged templates, Prisma Client escapes all variables.

In conclusion, the first three queries are safe from SQL Injection.

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏
